Wikipedia/assets/codemirror/codemirror-editTextSelection.js

```diff
@@ -0,0 +1,126 @@
+  }
+    
+  regexForLocatingSelectedTextWithPatternForSpace(patternForSpace, maxAdjacentWordsToUse) {
+    const replaceSpaceWith = (s, replacement) => s.replace(/\s+/g, replacement)
+
+    // Keep only the last 'maxAdjacentWordsToUse' words of 'textBeforeSelectedText'
+    const shouldKeepWordBeforeSelection = (e, i, a) => (a.length - i - 1) < maxAdjacentWordsToUse
+    // Keep only the first 'maxAdjacentWordsToUse' words of 'textAfterSelectedText'
+    const shouldKeepWordAfterSelection = (e, i) => i < maxAdjacentWordsToUse
+    const wordsBefore = this.textBeforeSelectedText.split(' ').filter(shouldKeepWordBeforeSelection).join(' ')
+    const wordsAfter = this.textAfterSelectedText.split(' ').filter(shouldKeepWordAfterSelection).join(' ')
+
+    const selectedTextPattern = replaceSpaceWith(this.selectedText, patternForSpace)
+    const textBeforeSelectedTextPattern = replaceSpaceWith(wordsBefore, patternForSpace)
+    const textAfterSelectedTextPattern = replaceSpaceWith(wordsAfter, patternForSpace)
+
+    // Attempt to locate wikitext selection based on the non-wikitext context strings above.
+    const atLeastOneNonWordOrOptionalStringPattern = '(?:\\W+|.*)'
+    const pattern = `(${textBeforeSelectedTextPattern.length > 0 ? '.*?' : ''}${textBeforeSelectedTextPattern}${atLeastOneNonWordOrOptionalStringPattern})(${selectedTextPattern})${atLeastOneNonWordOrOptionalStringPattern}${textAfterSelectedTextPattern}`
+    const regex = new RegExp(pattern, 's')
+
+    return regex
+  }
+}
+
+const wikitextRangeForSelectedAndAdjacentText = (selectedAndAdjacentText, wikitext) => {
+  const regex = selectedAndAdjacentText.regexForLocatingSelectedTextInWikitext(wikitext)
+  if (regex === null) {
+    return null
+  }
+  const match = wikitext.match(regex)
+  const matchedWikitextBeforeSelection = match[1]
+  const matchedWikitextSelection = match[2]
+  const wikitextRange = getWikitextRangeToSelect(matchedWikitextBeforeSelection, matchedWikitextSelection)
+  return wikitextRange
+}
+
+const getWikitextRangeToSelect = (wikitextBeforeSelection, wikitextSelection) => {
+  const wikitextBeforeSelectionLines = wikitextBeforeSelection.split('\n')
+  const startLine = wikitextBeforeSelectionLines.length - 1 
+  const startCh = wikitextBeforeSelectionLines.pop().length
+
+  const wikitextSelectionLines = wikitextSelection.split('\n')
+  const endLine = startLine + wikitextSelectionLines.length - 1
+  const endCh = wikitextSelectionLines.pop().length + (startLine === endLine ? startCh : 0)
+
+  let from = {line: startLine, ch: startCh}
+  let to = {line: endLine, ch: endCh}
+  
+  return {from, to}
+}
+
+const scrollToAndHighlightRange = (range, codemirror) => {
+  let marker = null
+  // Temporarily set selection so we can use existing `scrollCursorIntoViewIfNeeded` method to bring the selection on-screen.
+  codemirror.setSelection(range.from, range.to)
+  setTimeout(() => {
+    scrollCursorIntoViewIfNeeded(true)
```

Comment:
would it be appropriate to use `window.requestAnimationFrame` here instead of `setTimeout` or is this a different use case?

Reply:
good question! just checked - and the `removeAllRanges` call is a good use case for `window.requestAnimationFrame`

Reply:
fixed!

Reply:
based on @tonisevener's comments i was able to refactor this part be able to scroll to the location without using `setSelection` and `scrollCursorIntoView` - there's now a `scrollRangeIntoViewIfNeeded` which for this use case works better and fixes this 
https://github.com/wikimedia/wikipedia-ios/pull/2958#pullrequestreview-212389499 too.